### PR TITLE
#164798914 Implement User clap for article

### PR DESCRIPTION
**Pull request by samorahno**
#### Type of change
- [x] Endpoint to create clap
- [x] Write unit testing for create clap endpoint
#### How Has This Been Tested?
- All unit test passed
-  Successfully created new clap for user using postman
Screen Shot 2019-04-08 at 7 29 41 AM

#### Checklist:
- [x] Implement endpoint to create clap for article
- [x] All test passed
#### Story (PT)
[Delivers #164798913](https://www.pivotaltracker.com/story/show/164798913)

### CODE REVIEW

File: server/models/clap.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    articleId: {
+      type: DataTypes.UUID,
+      allowNull: false,
+    },
+    userId: {
+      type: DataTypes.UUID,
+      allowNull: false,
+    }
+  }, {});
+  Clap.associate = (models) => {};
```
Comment (davidokonji) [inline server/models/clap.js]: Good job, but there should be an association between claps and articles created, what do you think?

**Review by davidokonji — COMMENTED** [review]
Good work.

File: server/models/clap.js (hunk 1 above)
Comment (chukwudidu) [reply server/models/clap.js]: @davidokonji very true. To add to this, I think the route for this should be `api/v1/articles/:articleId/claps` following RESTful API convention?.

Comment (geekdanitek) [reply server/models/clap.js]: @davidokonji, the relationships between the claps and articles is out of scope for this pull request. The relationships should be for the person who will work on getting articles

**Review by geekdanitek — APPROVED** [review]

File: server/models/clap.js (hunk 1 above)
Comment (samorahno) [reply server/models/clap.js]: Thank you for the feedback @davidokonji, your point is valid, but I realised it is outside the scope of what I am working on.  That can be included for fetching all claps

Comment (samorahno) [reply server/models/clap.js]: @cupidcoder, I used this as a POST request, I feel the article id should be sent with the request body.

**Comment by samorahno** [comment]
> Hi @samorahno , I think export default should be used when you're exporting just one method. We should be using just 'export' for more than one method.

@BukkyOmo sequel to the convention that the team agreed to follow as regards module organization. I think altering this might affect other imports.

**Comment by Chiazokam** [comment]
Great work

File: server/models/clap.js (hunk 1 above)
Comment (BukkyOmo) [reply server/models/clap.js]: @geekdanitek how does creating a relationship between claps and articles have anything to do with fetching articles? Please do the needful.

**Comment by BukkyOmo** [comment]
> > Hi @samorahno , I think export default should be used when you're exporting just one method. We should be using just 'export' for more than one method.
> 
> @BukkyOmo sequel to the convention that the team agreed to follow as regards module organization. I think altering this might affect other imports.

Using export default to export more than one module is bad practice, export should be used instead. I understand that the group agreed to use export default, but you need to factor in cases when the codebase becomes large. It's already flagging errors in my code when I rebased. Moving forward, please let's use export. Thanks.

File: server/controllers/clap.controller.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import db from '../models';
+import Response from '../utils/response.util';
+
+const { Clap, Article } = db;
+/**
+ * Clap Controller
+ * @package clap
+ */
+class ClapController {
+  static async createClap(req, res) {
```
[the remark this answers is not in the document]
Comment (paradisekelechi) [reply server/controllers/clap.controller.js:16]: Hound issue persists

File: server/models/clap.js (hunk 1 above)
Comment (paradisekelechi) [reply server/models/clap.js]: @samorahno The recommendation made by @cupidcoder makes sense. `api/v1/articles/:articleId/claps` is a good representation. I do not want you to put the id in the body. The convention is that the id should be in the url as a parameter

File: server/controllers/clap.controller.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import Response from '../utils/response.util';
+
+const { Clap, Article } = db;
+/**
+ * Clap Controller
+ * @package clap
+ */
+class ClapController {
+  /**
+   * @description - Controller for create a clap endpoint
```
Comment (paradisekelechi) [inline server/controllers/clap.controller.js]: Grammatical error in this description

File: server/controllers/clap.controller.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+   * @param {object} req
+   * @param {object} res
+   * @returns {object} response
+   * @memberof createClap
```
Comment (paradisekelechi) [inline server/controllers/clap.controller.js]: Kindly remove the `memberof` annotation

File: server/controllers/clap.controller.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  */
+  static async createClap(req, res) {
+    const {
+      articleId
```
Comment (paradisekelechi) [inline server/controllers/clap.controller.js:18]: This should be in the params

File: server/controllers/clap.controller.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    } = req.body;
+    const { id } = req.user;
+    const successMessage = 'Clap successfully created';
```
Comment (paradisekelechi) [inline server/controllers/clap.controller.js]: You are using this variable once. No need to have it here. Use it directly where it should be used

File: server/controllers/clap.controller.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    const errorMessage = 'Clap was not created successfully';
```
Comment (paradisekelechi) [inline server/controllers/clap.controller.js]: You are using this variable once. No need to have it here. Use it directly where it should be used

File: server/controllers/clap.controller.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    const userCantClapTwice = await Clap.findOne({ where: { articleId, userId: id } });
```
Comment (paradisekelechi) [inline server/controllers/clap.controller.js]: The variable name should be improved upon. You can use something like `existingUserClap`

File: server/controllers/clap.controller.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    if (userCantClapTwice != null) return Response(res, 400, 'you cannot clap twice');
+    const cantClapSelf = await Article.findOne({ where: { userId: id }, as: 'user' });
```
Comment (paradisekelechi) [inline server/controllers/clap.controller.js]: The variable name should be improved upon. You can use something like `userOwnArticle`

File: server/models/clap.js (hunk 1 above)
Comment (samorahno) [reply server/models/clap.js]: Noted @paradisekelechi, this will be implemented right away

**Review by paradisekelechi — CHANGES_REQUESTED** [review]

**Pull request merged**